### ISSUE TITLE
Operand result type is returning as the negated QDM type

### ISSUE THREAD

**Issue by jackcmeyer** [cqframework/clinical_quality_language#115]
When using the following function in CQL, the operand for the function argument has a result type of the negated version of the QDM.

```
define function "func1"(AP "Assessment, Not Performed" ):
             true
```

This is inconsistent with other result type functionality.  For example, the following CQL definition returns the positive version of the QDM type

```
define "def1": 
              ["Assessment, Not Performed"]
```

Thoughts?

**Comment by brynrhodes** [comment]
So, this is an interesting one, and it really boils down to what you intended to do when you specified that the function takes a "Not Performed". Negation is accomplished through profiles in this way to enable the data access layer to communicate the profile information to the target system. If you look at the retrieve, you'll see the label of the profile is given as the template identifier in the retrieve.

But in the case of a function, we don't have that same need, and so if we dropped to the base type, we'd lose the fact that you declared the function to operator on "Not Performed". I'm inclined to say that since the profile is basically like a "marker interface", that we should leave it in the function and require the cast in order to use it as a negation, if that's what it is.

Does that make sense?

**Comment by JSRankins** [comment]
Hi Bryn. We agree that the "Not Performed" information should be kept. We are saying that no provision of the baseType for the associated profile label means that there cannot be direct type comparison of a list of returned "Not Performed" elements in a population to a function's argument type. We could say that this is intended, and that a function should not contain an argument that has a type of "Not Performed", but the CQL-based HQMF IG does not set this restriction. The other difficulty is that when the following CQL is created, it passes the current CQL-to-ELM validation:



```
define "def": 
    ["Assessment, Performed"] A
        where func(A)

define function "func"("AE" "Assessment, Not Performed" ): true
```

**Comment by brynrhodes** [comment]
I think it makes sense to still allow functions to be declared based on the "Not Performed", but it should require a "Not Performed". We were careful to exclude that from the return type of the retrieve to avoid implementers having to deal with Not Performed types, and we could still make a recommendation that Not Performed functions be disallowed, but I think that would make more sense as a program-level convention, rather than something in the IG.

You're example above though did uncover a problem with the type resolution in that the translator is automatically casting the Assessment, Performed to an Assessment, Not Performed because they are "compatible" (meaning they have the same attributes). That behavior is both surprising and wrong, and I'm fixing that issue.

**Comment by JSRankins** [comment]
Thanks Bryn. We'll check out the updated code. This appears to be created for CQL 1.2. When is the plan to push CQL 1.2 to master?

**Comment by brynrhodes** [comment]
Once CQL 1.2 has been published we'll switch the branch to master and formally release a 1.2 version. In the meantime, I've been pushing snapshots to sonatype. I just pushed another version with this and the other issue addressed (1.2.3-SNAPSHOT).
